ldif.go

```diff
@@ -0,0 +1,256 @@
+package ldap
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	// "os"
+	"strconv"
+)
+
+// A basic LDIF parser. This one does currently just support LDIFs like
+// they're generated by tools like ldapsearch(1) / slapcat(8). Change
+// records are not supported.
+type LDIF struct {
+	RelaxedParser bool // ignore parsing errors in a line
+	Entries       []*Entry
+	curLine       int
+	version       int
+	changeType    string
+}
+
+var CR byte = '\x0D'
+var LF byte = '\x0A'
+var SEP string = string([]byte{CR, LF})
+var Comment byte = '#'
+var SPACE byte = ' '
+var SPACES = string(SPACE)
+
+func (l *LDIF) newError(msg string) error {
+	return errors.New(fmt.Sprintf("error on line %d: %s\n", l.curLine, msg))
+}
+
+// Parses the LDIF, the caller is responsible for closing the io.Reader if that's
+// needed. This Parser may be called several times with different io.Readers to
+// combine the files.
+func (l *LDIF) Parse(r io.Reader) (err error) {
+	if r == nil {
+		return errors.New("No reader present")
+	}
+	reader := bufio.NewReader(r)
+	l.curLine = 0
+	l.changeType = ""
+	l.version = 0
+
+	var lines [][]byte
+	var line, nextLine []byte
+
+	for {
+		l.curLine++
+		nextLine, err = reader.ReadBytes(LF)
```

Comment:
if I'm reading the ldif rfc correctly, all the lines can be handled as strings. treating lines as strings throughout might clean up a lot of []byte casts